keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/actuation/ArtifactPromotionListener.kt

```diff
@@ -23,24 +23,25 @@ class ArtifactPromotionListener(
   @EventListener(ResourceDeltaResolved::class)
   fun onDeltaResolved(event: ResourceDeltaResolved) {
     val desired = event.desired
-    if (desired is ServerGroupSpec && desired.launchConfiguration.imageProvider is ArtifactImageProvider) {
-      val current = event.current as? ServerGroup
-      checkNotNull(current) {
-        "Current resource state is a ${event.current.javaClass.simpleName} when a ${ServerGroup::class.java.simpleName} was expected"
+    if (desired is ClusterSpec && desired.imageProvider is ArtifactImageProvider) {
+      val current = checkNotNull(event.current as? Set<ServerGroup>) {
+        "Current resource state is a ${event.current.javaClass.simpleName} when a collection of ${ServerGroup::class.java.simpleName} was expected"
       }
-      val appVersion = current.launchConfiguration.appVersion
-      val deliveryConfig = deliveryConfigRepository.deliveryConfigFor(event.resourceId)
-      val environment = deliveryConfigRepository.environmentFor(event.resourceId)
-      if (deliveryConfig == null || environment == null) {
-        log.warn("Resource ${event.resourceId} is not part of a delivery environment")
-      } else {
-        log.info("Marking {} as successfully deployed to {}'s {} environment", appVersion, deliveryConfig.name, environment.name)
-        artifactRepository.markAsSuccessfullyDeployedTo(
-          deliveryConfig,
-          desired.launchConfiguration.imageProvider.deliveryArtifact,
-          appVersion,
-          environment.name
-        )
+      current.forEach { serverGroup ->
```

Comment:
The current model can now include > 1 server group, so I'm just looping across them. Possibly we need to give this more thought.

Reply:
Raised #465 to track this. I think this PR is big enough already!